executor/common_linux.h

```diff
@@ -452,6 +463,74 @@ static void netdevsim_add(unsigned int addr, unsigned int port_count)
 	write_file("/sys/bus/netdevsim/new_device", buf);
 }
 
+#include <linux/devlink.h>
+#include <linux/genetlink.h>
+
+#define DEVLINK_FAMILY_NAME "devlink"
+
+#ifndef DEVLINK_ATTR_NETNS_FD
+#define DEVLINK_ATTR_NETNS_FD 137
+#endif
+
+static void netlink_devlink_netns_move(const char* bus_name, const char* dev_name, int netns_fd)
+{
+	struct genlmsghdr genlhdr;
+	struct nlattr* attr;
+	int sock, err, n;
+	uint16 id = 0;
+
+	sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_GENERIC);
+	if (sock == -1)
+		fail("socket(AF_NETLINK) failed\n");
+	memset(&genlhdr, 0, sizeof(genlhdr));
+	genlhdr.cmd = CTRL_CMD_GETFAMILY;
+	netlink_init(GENL_ID_CTRL, 0, &genlhdr, sizeof(genlhdr));
+	netlink_attr(CTRL_ATTR_FAMILY_NAME, DEVLINK_FAMILY_NAME, strlen(DEVLINK_FAMILY_NAME) + 1);
+	err = netlink_send_ext(sock, GENL_ID_CTRL, &n);
+	if (err) {
+		debug("netlink: failed to get devlink family id: %s\n", strerror(err));
+		goto error;
+	}
+	attr = (struct nlattr*)(nlmsg.buf + NLMSG_HDRLEN + NLMSG_ALIGN(sizeof(genlhdr)));
+	for (; (char*)attr < nlmsg.buf + n; attr = (struct nlattr*)((char*)attr + NLMSG_ALIGN(attr->nla_len))) {
+		if (attr->nla_type == CTRL_ATTR_FAMILY_ID) {
+			id = *(uint16*)(attr + 1);
+			break;
+		}
+	}
+	if (!id) {
+		debug("netlink: failed to parse message for devlink family id\n");
+		goto error;
+	}
+	recv(sock, nlmsg.buf, sizeof(nlmsg.buf), 0); /* recv ack */
+	memset(&genlhdr, 0, sizeof(genlhdr));
+	genlhdr.cmd = DEVLINK_CMD_RELOAD;
+	netlink_init(id, 0, &genlhdr, sizeof(genlhdr));
+	netlink_attr(DEVLINK_ATTR_BUS_NAME, bus_name, strlen(bus_name) + 1);
+	netlink_attr(DEVLINK_ATTR_DEV_NAME, dev_name, strlen(dev_name) + 1);
+	netlink_attr(DEVLINK_ATTR_NETNS_FD, &netns_fd, sizeof(netns_fd));
+	netlink_send(sock);
+error:
+	close(sock);
+}
+
+const int kInitNetNsFd = 239; // see kMaxFd
+
+static void initialize_devlink(void)
+{
+	int netns = open("/proc/self/ns/net", O_RDONLY);
+	if (netns == -1)
+		fail("open(/proc/self/ns/net) failed");
+	int ret = setns(kInitNetNsFd, 0);
+	if (ret == -1)
+		fail("set_ns(init_netns_fd) failed");
+	netlink_devlink_netns_move("pci", "0000:00:10.0", netns);
```

Comment:
This adds significant amount of code to every networking reproducer and some cost to setup of every test process. I think we need to make this a separate feature rather than bundle with SYZ_ENABLE_NETDEV, which is already too broad and does too many things.
Adding a separate feature will involve adding some code across a dozen of files, but it will allow to exclude this from reproducers and conditionally enable in test processes only if this device actually exists on the machine (and maybe of the right type, etc).
Check out how flag_enable_tun in executor works and these 2 commits for inspiration of how to add new features:
https://github.com/google/syzkaller/commit/b2f369e56e13dc135d57c53210ea7ab38b239e94
https://github.com/google/syzkaller/commit/8285069f89c9942f65ce760a8f0a5a12254bfeeb
In the end we need flag_enable_devlink_something in executor that will allow to skip initialization; SYZ_DEVLINK_SOMETHING macro that will allow to strip the code from reproducers and pkg/host.Features feature detected on startup.

Reply:
okay. Will fix.